tests/Http/Handlers/ClientHandlerTest.php

```diff
@@ -0,0 +1,38 @@
+<?php
+
+declare(strict_types=1);
+
+namespace HelpScout\Api\Tests\Http\Handlers;
+
+use GuzzleHttp\Exception\RequestException;
+use HelpScout\Api\Tests\ApiClientIntegrationTestCase;
+
+class ClientHandlerTest extends ApiClientIntegrationTestCase
+{
+    public function testRequestExceptionThrownWhenBadRequest()
+    {
+        $this->expectException(RequestException::class);
+
+        $this->stubResponse($this->getResponse(400, json_encode([])));
+
+        $this->client->customers()->get(1);
+    }
+
+    public function testRequestExceptionThrownWhenAccessDenied()
+    {
+        $this->expectException(RequestException::class);
+
+        $this->stubResponse($this->getResponse(403, json_encode([])));
+
+        $this->client->customers()->get(1);
+    }
```

Comment:
I think we just add 1 test here that doesn't throw an exception if the response is `401` and it'll be good to go

Reply:
While digging in and writing this test I found out that this testable scenario will only occur if **the Authenciationhandler is _not_ added to the stack** [here](https://github.com/helpscout/helpscout-api-php/blob/27395a069c02a01f35957b34deb5175112746bda/src/Http/RestClientBuilder.php#L94). If I'm correct, as long as the AuthenticationHandler is pushed to the Handler stack, then a `401` would return an exception and not the next request. 

Example test:

```php
public function testRequestExceptionNotThrownWhenNotAuthorized()
    {
        $this->mockHandler = new MockHandler();
        $handler = HandlerStack::create($this->mockHandler);

        $handler->push(Middleware::history($this->history));
        // AuthenticationHandler not pushed to Handler stack
        $handler->push(new ClientErrorHandler());
        $handler->push(new RateLimitHandler());
        $handler->push(new ValidationHandler());

        $client = new Client(['handler' => $handler, 'http_errors' => false]);
        $this->authenticator = new Authenticator($client);
        $this->authenticator->setAccessToken('abc123');

        $this->client = new ApiClient(
            new RestClient($client, $this->authenticator)
        );
        $this->stubResponse($this->getResponse(401, json_encode([])));
        $customer = $this->client->customers()->get(1);
        $this->assertInstanceOf(Customer::class, $customer);
    }
```

Reply:
We could reduce the complexity of the test by only testing the handler itself, avoiding us having to build the several `Client` and `Handler` objects.  I haven't executed this so this isn't a complete test, but here's an example of what I'm thinking:

```
$expectedResponse = Mockery::mock(ResponseInterface::class);
$handler = new AuthenticationHandler();
$response = $handler(function () use ($expectedResponse) {
    $promise = new Promise(function () use (&$promise, $expectedResponse) {
        $promise->resolve(expectedResponse);
    });
    return $promise;
});
$this->assertEquals($expectedResponse, $response);
```

There's docs on Guzzle's promises at https://github.com/guzzle/promises.

Reply:
Very nice, 👏

Reply:
I was fussing with this for a bit, but was able to trim down this test and still have it passing. Let me know what you think.

```php
public function testRequestExceptionNotThrownWhenNotAuthorized()
    {
        $this->mockHandler = new MockHandler();
        $handler           = HandlerStack::create($this->mockHandler);

        $client              = new Client(['handler' => $handler]);
        $this->authenticator = new Authenticator($client);
        $this->authenticator->setAccessToken('abc123');

        $this->client = new ApiClient(
            new RestClient($client, $this->authenticator)
        );

        $this->stubResponse($this->getResponse(401, json_encode([])));

        $this->assertInstanceOf(
            Customer::class,
            $this->client->customers()->get(1)
        );
    }
```

Reply:
Looks great, well done!